### PR TITLE
Update compliments with support for cron type date/time for selections, addition to just date.

### PR DESCRIPTION
**Pull request by sdetweil**
> - What does the pull request accomplish? Use a list if needed.

this change allows uses to configure date/time events for compliments..
also linked site that will build the cron entry..

and example was Happy hour in a pub, on fri/sat between 5 and 7 pm. 

or just after midnight on Halloween (Boooooo!)

I also added testcases for #3478 

(and added support for this in MMM-Config), with a custom, drop down selection list of the types.. )

|  if this is approved I will update the module doc

### CODE REVIEW

**Comment by sdetweil** [comment]
the same failing 2 tests run correctly in the electron test suite..  what is the problem here?

copied from the electron test suite and added the get.document() call before the expect.   but only these  two fail.
like the date isn't set

**Comment by sdetweil** [comment]
ok, the problem here is the startApplication() helper in e2e does not accept dates for the test..  the electron version does accept date

so we cannot run date based testcases in e2e .. the electron tests DO pass

**Comment by sdetweil** [comment]
the weather testcase depends on a weather provider responding to the special value in the config 
```js
		{
			module: "weather",
			position: "bottom_bar",
			config: {
				location: "Munich",
				mockData: '"#####WEATHERDATA#####"'
			}
		}
```
but the default provider  does not send any notification..

**Comment by khassel** [comment]
untouched, but you changed compliment sources

**Comment by sdetweil** [comment]
yes, but that testcase is in weather.. not in compliments
and that part of compliments wasnt changed. it depends on the notification

```js
	describe("Compliments Integration", () => {
		beforeAll(async () => {
			await weatherFunc.startApp("tests/configs/modules/weather/currentweather_compliments.js", {});
		});

		it("should render a compliment based on the current weather", async () => {
			await expect(weatherFunc.getText(".compliments .module-content span", "snow")).resolves.toBe(true);
		});
	});
```
i tested and the compliments notification_received is not triggered.. 

in compliments.js 
```js
		// Add compliments based on weather
		if (this.currentWeatherType in this.config.compliments) {
			Array.prototype.push.apply(compliments, this.config.compliments[this.currentWeatherType]);
			// if the predefine list doesn't include it (yet)             // i added these two lines
			if (!this.pre_defined_types.includes(this.currentWeatherType)) {
				// add it
				this.pre_defined_types.push(this.currentWeatherType);
			}
		}

		// Add compliments for anytime
		Array.prototype.push.apply(compliments, this.config.compliments.anytime);
```

the config for the weather_compliments testcase 
```js
let config = {
	modules: [
		{
			module: "compliments",
			position: "top_bar",
			config: {
				compliments: {
					snow: ["snow"]
				},
				updateInterval: 3000
			}
		},
		{
			module: "weather",
			position: "bottom_bar",
			config: {
				location: "Munich",
				mockData: '"#####WEATHERDATA#####"'
			}
		}
	]
};
```
doesn't have date based, and doesn't have normal ,, so depends on the notification, which doesn't arrive... 
```js
	// Override notification handler.
	notificationReceived (notification, payload, sender) {
		if (notification === "CURRENTWEATHER_TYPE") {
			this.currentWeatherType = payload.type;
		}
	}
```
I forced this.currentWeatherType to 'snow' with a compliment under "snow" and it appeared.. so the compliments module is not broken

**Comment by khassel** [comment]
with old `compliments.js`:

```bash
node@cb3498f2205a:/opt/magic_mirror$ npx jest tests/e2e/modules/weather_current_spec.j --force-exit
 PASS   e2e  tests/e2e/modules/weather_current_spec.js (21.22 s)
  Weather module
    Current weather
      Default configuration
        ✓ should render wind speed and wind direction (814 ms)
        ✓ should render temperature with icon (201 ms)
        ✓ should render feels like temperature (200 ms)
        ✓ should render humidity next to feels-like (201 ms)
    Compliments Integration
      ✓ should render a compliment based on the current weather (5315 ms)
    Configuration Options
      ✓ should render windUnits in beaufort (304 ms)
      ✓ should render windDirection with an arrow (206 ms)
      ✓ should render humidity next to wind (204 ms)
      ✓ should render degreeLabel for temp (203 ms)
      ✓ should render degreeLabel for feels like (202 ms)
    Current weather with imperial units
      ✓ should render wind in imperial units (306 ms)
      ✓ should render temperatures in fahrenheit (201 ms)
      ✓ should render 'feels like' in fahrenheit (203 ms)

Test Suites: 1 passed, 1 total
Tests:       13 passed, 13 total
Snapshots:   0 total
Time:        21.85 s
Ran all test suites matching /tests\/e2e\/modules\/weather_current_spec.j/i.
```

with `compliments.js` of this PR:
```bash
node@cb3498f2205a:/opt/magic_mirror$ npx jest tests/e2e/modules/weather_current_spec.j --force-exit
 FAIL   e2e  tests/e2e/modules/weather_current_spec.js (6.252 s)
  Weather module
    Current weather
      Default configuration
        ✓ should render wind speed and wind direction (305 ms)
        ✓ should render temperature with icon (201 ms)
        ✓ should render feels like temperature (202 ms)
        ✓ should render humidity next to feels-like (202 ms)
    Compliments Integration
      ✕ should render a compliment based on the current weather (212 ms)
    Configuration Options
      ✓ should render windUnits in beaufort (308 ms)
      ✓ should render windDirection with an arrow (206 ms)
      ✓ should render humidity next to wind (200 ms)
      ✓ should render degreeLabel for temp (203 ms)
      ✓ should render degreeLabel for feels like (203 ms)
    Current weather with imperial units
      ✓ should render wind in imperial units (300 ms)
      ✓ should render temperatures in fahrenheit (202 ms)
      ✓ should render 'feels like' in fahrenheit (200 ms)

  ● Weather module › Compliments Integration › should render a compliment based on the current weather

    expect(received).resolves.toBe()

    Received promise rejected instead of resolved
    Rejected to value: [Error: expect(received).toBe(expected) // Object.is equality·
    Expected: "snow"
    Received: ""]

      39 |
      40 |              it("should render a compliment based on the current weather", async () => {
    > 41 |                      await expect(weatherFunc.getText(".compliments .module-content span", "snow")).resolves.toBe(true);
         |                            ^
      42 |              });
      43 |      });
      44 |

      at expect (node_modules/expect/build/index.js:113:15)
      at Object.expect (tests/e2e/modules/weather_current_spec.js:41:10)

Test Suites: 1 failed, 1 total
Tests:       1 failed, 12 passed, 13 total
Snapshots:   0 total
Time:        6.297 s
Ran all test suites matching /tests\/e2e\/modules\/weather_current_spec.j/i.
```

**Comment by sdetweil** [comment]
looking.. how do you debug the code here?

**Comment by khassel** [comment]
> looking.. how do you debug the code here?

within the test? This is difficult, AFAIR only option is `console.dir()`

File: modules/default/compliments/compliments.js [hunk 1]
```diff
@@ -1,3 +1,9 @@
+/* MagicMirror²
```
Comment (khassel) [inline modules/default/compliments/compliments.js:1]: why are the header comments back? We removed them from all files ...

File: tests/configs/modules/compliments/compliments_specialDayUnique_false.js [hunk 2]
```diff
@@ -4,14 +4,15 @@ let config = {
 			module: "compliments",
 			position: "middle_center",
 			config: {
+				updateInterval: 1000 * 5, // Update every 5 secs
```
Comment (khassel) [inline tests/configs/modules/compliments/compliments_specialDayUnique_false.js:7]: do we need updateInterval here? I think not.

File: tests/configs/modules/compliments/compliments_specialDayUnique_true.js [hunk 3]
```diff
@@ -4,14 +4,15 @@ let config = {
 			module: "compliments",
 			position: "middle_center",
 			config: {
+				updateInterval: 1000 * 5, // Update every 5 secs
```
Comment (khassel) [inline tests/configs/modules/compliments/compliments_specialDayUnique_true.js:7]: same as above

**Comment by sdetweil** [comment]
updateInterval from testcases for #3471, I didn't check that ,, fixed
header.. oops I started with a prior version,, didn't catch that.. fixed

File: tests/configs/modules/compliments/compliments_cron_entry.js [hunk 4]
```diff
@@ -0,0 +1,19 @@
+let config = {
+	modules: [
+		{
+			module: "compliments",
+			position: "middle_center",
+			config: {
+				updateInterval: 1000 * 5, // Update every 5 secs
```
Comment (khassel) [inline tests/configs/modules/compliments/compliments_cron_entry.js:7]: did not see this one at first review, can you please remove this updateInterval too ...

**Comment by sdetweil** [comment]
done

File: tests/e2e/modules/compliments_spec.js [hunk 5]
```diff
@@ -1,81 +0,0 @@
-const helpers = require("../helpers/global-setup");
```
Comment (khassel) [inline tests/e2e/modules/compliments_spec.js:1]: why did you delete this file? I don't see that all tests from here moved to electron section so we lost these tests.

**Comment by sdetweil** [comment]
because I had to fix the prior testcases, it made me realize there a cron anytime as well.. so I added that to e2e configs and tests

**Comment by sdetweil** [comment]
@rejas  please review. I'd love  your feedback

File: modules/default/compliments/compliments.js [hunk 6]
```diff
@@ -38,11 +40,40 @@ Module.register("compliments", {
...
+	 */
+	getSecondsUntilNextCronRun (cronExpression, timestamp = new Date()) {
+		// Required for seconds precision
+		const adjustedTimestamp = new Date(timestamp.getTime() - 1000);
 
-		// Schedule update timer.
-		setInterval(() => {
-			this.updateDom(this.config.fadeSpeed);
-		}, this.config.updateInterval);
+		// https://www.npmjs.com/package/croner
```
Comment (rejas) [inline modules/default/compliments/compliments.js:70]: this isnt needed I guess?

File: modules/default/compliments/compliments.js [hunk 7]
```diff
@@ -38,11 +40,40 @@ Module.register("compliments", {
...
+		/*eslint-disable-next-line*/
```
Comment (rejas) [inline modules/default/compliments/compliments.js:71]: why this disable?

File: modules/default/compliments/compliments.js [hunk 8]
```diff
@@ -38,11 +40,40 @@ Module.register("compliments", {
...
+		const cronJob = new Cron(cronExpression);
+		const nextRunTime = cronJob.nextRun(adjustedTimestamp);
+
+		let secondsDelta = (nextRunTime - adjustedTimestamp) / 1000;
```
Comment (rejas) [inline modules/default/compliments/compliments.js:75]: can be a const?

File: modules/default/compliments/compliments.js [hunk 9]
```diff
@@ -51,7 +82,7 @@ Module.register("compliments", {
 	 * @returns {number} a random index of given array
 	 */
 	randomIndex (compliments) {
-		if (compliments.length <= 1) {
+		if (compliments.length === 1) {
```
Comment (rejas) [inline modules/default/compliments/compliments.js:85]: why this change?

File: modules/default/compliments/compliments.js [hunk 10]
```diff
@@ -91,20 +123,49 @@ Module.register("compliments", {
...
-		for (let entry in this.config.compliments) {
-			if (new RegExp(entry).test(date)) {
-				// Only display compliments configured for the day if specialDayUnique is set to true
-				if (this.config.specialDayUnique) {
-					compliments.length = 0;
-				}
-				Array.prototype.push.apply(compliments, this.config.compliments[entry]);
+		for (let entry of temp_list) {
+			// check if this could be a cron type entry
+			if (entry.includes(" ")) {
```
Comment (rejas) [inline modules/default/compliments/compliments.js:146]: we got a m.includes(" ") some lines above. Maybe just make this a function
checkCronEntry 
that does this. Make it clearer and gets rid of the comments

**Review by rejas — CHANGES_REQUESTED** [review]

File: modules/default/compliments/compliments.js (hunk 7 above)
Comment (sdetweil) [reply modules/default/compliments/compliments.js:71]: because the linter fails that line cause there is no matching function definition (require) as the script is loaded thru getScripts

File: modules/default/compliments/compliments.js (hunk 8 above)
Comment (sdetweil) [reply modules/default/compliments/compliments.js:75]: i'm still old school.. can't be a const if its set my code..  const is a literal.  (imho)

File: modules/default/compliments/compliments.js (hunk 9 above)
Comment (sdetweil) [reply modules/default/compliments/compliments.js:85]: i sure don't remember changing that

File: modules/default/compliments/compliments.js (hunk 6 above)
Comment (sdetweil) [reply modules/default/compliments/compliments.js:70]: eh?  where is line 43 in the new source?

File: modules/default/compliments/compliments.js (hunk 10 above)
Comment (sdetweil) [reply modules/default/compliments/compliments.js:146]: ok

File: modules/default/compliments/compliments.js (hunk 6 above)
Comment (sdetweil) [reply modules/default/compliments/compliments.js:76]: that code is at line 55

File: modules/default/compliments/compliments.js (hunk 8 above)
Comment (sdetweil) [reply modules/default/compliments/compliments.js]: fixed

File: modules/default/compliments/compliments.js (hunk 7 above)
Comment (sdetweil) [reply modules/default/compliments/compliments.js:77]:
```sh
/home/sam/MagicMirror/modules/default/compliments/compliments.js
  78:23  error  'Cron' is not defined  no-undef
```

Comment (rejas) [reply modules/default/compliments/compliments.js:77]: https://gitkraken.dev/link/drafts/c673b0d9-8e28-4640-8fe5-7e22673262d4?type=suggested_pr_change&prEntityId=WyJnaXRodWIiLCJwciIsIjEiLCIiLCJQUl9rd0RPQVFNOHU4NXpiTl8tIl0=

Comment (rejas) [reply modules/default/compliments/compliments.js:77]: Then add a global to the top of the file (like we do for other libs):

/* global Cron */

Comment (sdetweil) [reply modules/default/compliments/compliments.js]: ok, that makes no sense.. but it worked..

**Review by rejas — APPROVED** [review]
